draft-trammell-taps-post-sockets.md

```diff
@@ -318,40 +318,40 @@ interaction patterns, defined in the subsections below.
    applications. A Responder's receiver gets a Message, as well as a Source to
    send replies to. Responders cannot be forked, and need not accept forks.
 
- - Stream:
-   A Message Carrier may be irreversibly morphed into a Stream, in order to provide
-   a strictly ordered, reliable service as with SOCK_STREAM. Morphing a Message
-   Carrier into a Stream should return a "file-like object" as appropriate for the
-   platform implementing the API. Typically, both ends of a communication using a
-   stream service will morph their respective Message Carriers independently before
-   sending any Messages.
-
-   Writing a byte to a Stream will cause it to be received by the remote, in
-   order, or will cause an error condition and termination of the stream if the
-   byte cannot be delivered. Due to the strong sequential dependence on a stream,
-   streams must always be reliable and ordered. A Message Carrier may only be
-   morphed to a Stream if it uses transport protocol stack that provides
-   reliable, ordered service, and only before it is used to send a Message.
-
 ## Message
 
-A Message is an atomic unit of communication between applications. A Message
-that cannot be delivered in its entirety within the constraints of the network
-connectivity and the requirements of the application is not delivered at all.
-
-Messages can represent both relatively small structures, such as requests in a
+A Message is the unit of communication between applications. Messages
+can represent both relatively small structures, such as requests in a
 request/response protocol such as HTTP; as well as relatively large
 structures, such as files of arbitrary size in a filesystem.
 
 In the general case, there is no mapping between a Message and packets sent by
 the underlying protocol stack on the wire: the transport protocol may freely
 segment messages and/or combine messages into packets. However, a message may be
-marked as immediate, which will cause it to be sent in a single packet, if it
-will fit.
-
-This implies that both the sending and receiving endpoint, whether in the
-application layer or the transport layer, must guarantee storage for the full
-size of an Message.
+marked as immediate, which will cause it to be sent in a single packet when
+possible.
+
+Messages may be sent and received either atomically, or non-atomically. All
+protocols must support non-atomic Messages, but only protocols that have
+fixed maximum lengths may allow clients to treat Messages as atomic.
+
```

Comment:
I agree that 'atomic' is probably not the best name. Also I would rather see atomic message as the default while non-atomic as the exception. Of course this is not the cause if you see TCP as the main protocol to be used, but normally it's very rare that you have a raw tcp with streaming semantics while usually there is message framing implemented on top. Your text seems to rather take non-atomic as the default that’s always supported by all protocols, however, at which point in time you tell the app that data is ready to be consumed is really an API-only question and basically independent of the protocol. 

Further, and maybe I’m reading a bit between the lines, you also seem to assume a synchronous pull scheme is used with non-atomic message(s). This might makes sense because one long message actually means that the transport layer has now clue when there is a good chunk of data to notify the application about. However, there is another data unit the transport knows about, which is packets. This is rather arbitrary regarding the content of these packets but it could still be used to implement an asynchronous push interface to keep either the notifications low or make sure you receive data as quickly as possible ("notify me very X successfully received in-order packets“). In any case any assumption about using a pull or push mechanism needs to be spelled out more explicitly.